src/checker.ml

```diff
@@ -126,43 +131,82 @@ let[@inline] ensure_valid_leaf_ptr (mem: mem) (leaf_ptr: leaf_ptr) : unit =
   | Some (Leaf _) -> ()
   | _ -> Ligo.failwith error_InvalidLeafPtr
 
-let[@inline] entrypoint_create_burrow (state, (burrow_no, delegate_opt): checker * (Ligo.nat * Ligo.key_hash option)) =
+(* TODO: if no tez is allowed to be given anywhere, it might be a better idea
+ * to check it once in checkerMain.ml instead of in each entrypoint. *)
+
+let[@inline] entrypoint_create_burrow (state, (burrow_no, delegate_opt, tok): checker * (Ligo.nat * Ligo.key_hash option * tok)) =
   assert_checker_invariants state;
+  let _ = ensure_no_tez_given () in
   let burrow_id = (!Ligo.Tezos.sender, burrow_no) in
   let () = if Ligo.Big_map.mem burrow_id state.burrows
     then Ligo.failwith error_BurrowAlreadyExists
     else () in
-  let op, burrow_address =
+  (* OP 1: Originate a new burrow *)
+  let origination_op, burrow_address =
     LigoOp.Tezos.burrow_create_contract
       (fun (p, storage : burrow_parameter * burrow_storage) ->
          if !Ligo.Tezos.sender <> storage.checker_address then
            (Ligo.failwith (Ligo.int_from_literal "-1") : LigoOp.operation list * burrow_storage)
+         else if !Ligo.Tezos.amount <> Ligo.tez_from_literal "0mutez" then
+           (Ligo.failwith (Ligo.int_from_literal "-2") : LigoOp.operation list * burrow_storage)
          else
            match p with
            | BurrowSetDelegate kho ->
-             ([LigoOp.Tezos.set_delegate kho], storage)
-           | BurrowStoreTez ->
-             (([] : LigoOp.operation list), storage)
-           | BurrowSendTezTo p ->
-             let (tez, addr) = p in
-             let op = match (LigoOp.Tezos.get_contract_opt addr : unit Ligo.contract option) with
-               | Some c -> LigoOp.Tezos.unit_transaction () tez c
-               | None -> (Ligo.failwith (Ligo.int_from_literal "-2") : LigoOp.operation) in
+             (* TODO: this deviates slightly from the design in the issue;
+              * I think the case analysis should happen in burrow code, and
+              * not in checker code. *)
+             (* TODO: Also, I am not happy that an FA2 contract could "pretend"
+              * to have a set_delegate entrypoint and succeed here. *)
```

Comment:
I guess it's not that big of an issue after all :thinking:


I considered whether delegation should happen from the TezWrapper side (would also significantly lower the gas costs, not having to go through checker), but that can't happen easily: (a) the burrow contracts only listen to checker, and (b) only checker knows their address.

I considered exposing the burrow address for a user via an offline view to address (b), and changing the current logic to circumvent (a). There's still the issue of how much data would we need to keep in the vault's storage though to change the permission checking.

Not sure if any of that will work, but it might be worth considering it as an improvement after we merge this PR.